src/nc_server.c

```diff
@@ -266,31 +267,37 @@ static void
 server_failure(struct context *ctx, struct server *server)
 {
     struct server_pool *pool = server->owner;
-    int64_t now, next;
+    int64_t now;
     rstatus_t status;
+    bool is_reconnect;
 
     if (!pool->auto_eject_hosts) {
         return;
     }
 
-    server->failure_count++;
-
     log_debug(LOG_VERB, "server '%.*s' failure count %"PRIu32" limit %"PRIu32,
               server->pname.len, server->pname.data, server->failure_count,
               pool->server_failure_limit);
 
-    if (server->failure_count < pool->server_failure_limit) {
+    now = nc_usec_now();
+    if (now < 0) {
         return;
     }
 
-    now = nc_usec_now();
-    if (now < 0) {
+    server->next_retry = now + pool->server_retry_timeout;
+    server->failure_count++;
+    is_reconnect = (server->fail != FAIL_STATUS_NORMAL) ? true : false;
+
+    if (is_reconnect) {
+        add_failed_server(ctx, server);
```

Comment:
Would it make more sense to remove server_failure_limit entirely (i.e. same as always 1)? Is it still needed with a heartbeat patch?

Also, I see that add_failed_server will set server->fail = FAIL_STATUS_ERR_TRY_HEARTBEAT and server->next_retry is also set.

This branch won't call server_pool_run to update the distribution(e.g. ketama), but if something unrelated updates the ketama distribution (server_pool_run) (e.g. a different memcache server in the same pool changes state), then the server would be ejected from the pool.